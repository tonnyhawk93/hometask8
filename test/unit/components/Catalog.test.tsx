import { it, expect, describe} from '@jest/globals';
import { FakeApi, FakeCartApi } from './fakeApi/fakeApi';
import React from 'react';
import { render, screen, waitForElementToBeRemoved, waitFor} from '@testing-library/react';
import event from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { Catalog } from '../../../src/client/pages/Catalog';
import { ProductDetails } from '../../../src/client/components/ProductDetails';
import { initStore } from '../../../src/client/store';
import { BrowserRouter } from 'react-router-dom';

describe('Каталог', () => {
    it('В каталоге должны отображаться товары, список которых приходит с сервера', async () => {
        const products = [
            {
                id: 0,
                name: 'Tasty Pants',
                description: 'Andy shoes are designed to keeping in mind durability as well as trends, the most stylish range of shoes & sandals',
                price: 302,
                color: 'Purple',
                material: 'Metal',
            },
            {
                id: 1,
                name: 'Tasty Towels',
                description: "Boston's most advanced compression wear technology increases muscle oxygenation, stabilizes active muscles",
                price: 614,
                color: 'Red',
                material: 'Soft',
            },
            {
                id: 2,
                name: 'Handcrafted Computer',
                description: 'The beautiful range of Apple Naturalé that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients',
                price: 131,
                color: 'Ivory',
                material: 'Rubber',
            },
            {
                id: 3,
                name: 'Licensed Shirt',
                description: 'Carbonite web goalkeeper gloves are ergonomically designed to give easy fit',
                price: 78,
                color: 'Teal',
                material: 'Frozen',
            },
            {
                id: 4,
                name: 'Handcrafted Mouse',
                description: 'The Football Is Good For Training And Recreational Purposes',
                price: 225,
                color: 'Pink',
                material: 'Steel',
            },
        ]
        const api = new FakeApi('test-basename', products);
        const cartApi = new FakeCartApi({});
        const store = initStore(api, cartApi);

        const catalog = (
            <BrowserRouter>
                <Provider store={store}>
                    <Catalog />
                </Provider>
            </BrowserRouter>
        );
        const { getAllByTestId } = render(catalog);
        await waitForElementToBeRemoved(() => screen.getByText(/loading/i));

        for(let product of products) {
            getAllByTestId(product.id)
        }
    })

    it('Для каждого товара в каталоге отображается название, цена и ссылка на страницу с подробной информацией о товаре', async () => {
        const products = [
            {
                id: 0,
                name: 'Tasty Pants',
                description: 'Andy shoes are designed to keeping in mind durability as well as trends, the most stylish range of shoes & sandals',
                price: 302,
                color: 'Purple',
                material: 'Metal',
            }
        ]
        const api = new FakeApi('test-basename', products);
        const cartApi = new FakeCartApi({});
        const store = initStore(api, cartApi);

        const catalog = (
            <BrowserRouter>
                <Provider store={store}>
                    <Catalog />
                </Provider>
            </BrowserRouter>
        );
        const { getAllByTestId } = render(catalog);
        await waitForElementToBeRemoved(() => screen.getByText(/loading/i));
        expect(screen.getByText('Tasty Pants')).toBeTruthy();
        expect(screen.getByText('$302')).toBeTruthy();
        expect(screen.getByText(/details/i).getAttribute('href')).toEqual('/catalog/0')
    })
    
    it('На странице с подробной информацией отображаются: название товара, его описание, цена, цвет, материал и кнопка "добавить в корзину"', async () => {
        const products = [
            {
                id: 0,
                name: 'Tasty Pants',
                description: 'test-descr',
                price: 302,
                color: 'Purple',
                material: 'Metal',
            }
        ]
        const api = new FakeApi('test-basename', products);
        const cartApi = new FakeCartApi({});
        const store = initStore(api, cartApi);
        const catalog = (
            <BrowserRouter>
                <Provider store={store}>
                    <ProductDetails product = {products[0]}/>
                </Provider>
            </BrowserRouter>
        );
        const { container } = render(catalog);
        expect(screen.getByText('Tasty Pants')).toBeTruthy();
        expect(screen.getByText('$302')).toBeTruthy();
        expect(screen.getByText('test-descr')).toBeTruthy();
        expect(screen.getByText('Purple')).toBeTruthy();
        expect(screen.getByText('Metal')).toBeTruthy();
        expect(screen.getByRole('button', { name: /add to cart/i })).toBeTruthy();
    })
})

