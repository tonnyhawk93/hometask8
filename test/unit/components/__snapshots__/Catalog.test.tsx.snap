// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Каталог На ширине меньше 576px навигационное меню должно скрываться за "гамбургер" 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Application"
      >
        <nav
          class="navbar navbar-expand-sm navbar-light bg-light"
        >
          <div
            class="container"
          >
            <a
              class="Application-Brand navbar-brand"
              href="/"
            >
              Example store
            </a>
            <button
              aria-label="Toggle navigation"
              class="Application-Toggler navbar-toggler"
            >
              <span
                class="navbar-toggler-icon"
              />
            </button>
            <div
              class="Application-Menu collapse navbar-collapse"
            >
              <div
                class="navbar-nav"
              >
                <a
                  class="nav-link"
                  href="/catalog"
                >
                  Catalog
                </a>
                <a
                  class="nav-link"
                  href="/delivery"
                >
                  Delivery
                </a>
                <a
                  class="nav-link"
                  href="/contacts"
                >
                  Contacts
                </a>
                <a
                  class="nav-link"
                  href="/cart"
                >
                  Cart
                </a>
              </div>
            </div>
          </div>
        </nav>
        <div
          class="container pt-4"
        >
          <div
            class="Home"
          >
            <div
              class="row"
            >
              <div
                class="col bg-secondary text-white py-4 bg-opacity-75"
              >
                <p
                  class="display-3"
                >
                  Welcome to Example store!
                </p>
                <p
                  class="lead"
                >
                  Culpa perspiciatis corporis facilis fugit similique
                </p>
                <p
                  class="lead"
                >
                  Cum aut ut eveniet rem cupiditate natus veritatis quia
                </p>
              </div>
            </div>
            <div
              class="row mb-4"
            >
              <div
                class="col-12 col-md-4 bg-light py-3"
              >
                <h1>
                  Quickly
                </h1>
                <p
                  class="lead"
                >
                  Odio aut assumenda ipsam amet reprehenderit. Perspiciatis qui molestiae qui tempora quisquam
                </p>
              </div>
              <div
                class="col-12 col-md-4 bg-light py-3"
              >
                <h1>
                  Qualitatively
                </h1>
                <p
                  class="lead"
                >
                  Ut nisi distinctio est non voluptatem. Odio aut assumenda ipsam amet reprehenderit
                </p>
              </div>
              <div
                class="col-12 col-md-4 bg-light py-3"
              >
                <h1>
                  Inexpensive
                </h1>
                <p
                  class="lead"
                >
                  Perspiciatis qui molestiae qui tempora quisquam. Ut nisi distinctio est non voluptatem
                </p>
              </div>
            </div>
            <div
              class="row mb-4"
            >
              <div
                class="col-12py-3"
              >
                <p>
                  Sed voluptatum quis voluptates laudantium incidunt laudantium. Illo non quos eos vel ipsa. Explicabo itaque est optio neque rerum provident enim qui sed. Corrupti commodi voluptatem vero soluta hic.
                </p>
                <p>
                  Modi corporis consectetur aliquid sit cum tenetur enim. Sed voluptatum quis voluptates laudantium incidunt laudantium. Illo non quos eos vel ipsa. Explicabo itaque est optio neque rerum provident enim qui sed. Corrupti commodi voluptatem vero soluta hic.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Application"
    >
      <nav
        class="navbar navbar-expand-sm navbar-light bg-light"
      >
        <div
          class="container"
        >
          <a
            class="Application-Brand navbar-brand"
            href="/"
          >
            Example store
          </a>
          <button
            aria-label="Toggle navigation"
            class="Application-Toggler navbar-toggler"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="Application-Menu collapse navbar-collapse"
          >
            <div
              class="navbar-nav"
            >
              <a
                class="nav-link"
                href="/catalog"
              >
                Catalog
              </a>
              <a
                class="nav-link"
                href="/delivery"
              >
                Delivery
              </a>
              <a
                class="nav-link"
                href="/contacts"
              >
                Contacts
              </a>
              <a
                class="nav-link"
                href="/cart"
              >
                Cart
              </a>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="container pt-4"
      >
        <div
          class="Home"
        >
          <div
            class="row"
          >
            <div
              class="col bg-secondary text-white py-4 bg-opacity-75"
            >
              <p
                class="display-3"
              >
                Welcome to Example store!
              </p>
              <p
                class="lead"
              >
                Culpa perspiciatis corporis facilis fugit similique
              </p>
              <p
                class="lead"
              >
                Cum aut ut eveniet rem cupiditate natus veritatis quia
              </p>
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <div
              class="col-12 col-md-4 bg-light py-3"
            >
              <h1>
                Quickly
              </h1>
              <p
                class="lead"
              >
                Odio aut assumenda ipsam amet reprehenderit. Perspiciatis qui molestiae qui tempora quisquam
              </p>
            </div>
            <div
              class="col-12 col-md-4 bg-light py-3"
            >
              <h1>
                Qualitatively
              </h1>
              <p
                class="lead"
              >
                Ut nisi distinctio est non voluptatem. Odio aut assumenda ipsam amet reprehenderit
              </p>
            </div>
            <div
              class="col-12 col-md-4 bg-light py-3"
            >
              <h1>
                Inexpensive
              </h1>
              <p
                class="lead"
              >
                Perspiciatis qui molestiae qui tempora quisquam. Ut nisi distinctio est non voluptatem
              </p>
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <div
              class="col-12py-3"
            >
              <p>
                Sed voluptatum quis voluptates laudantium incidunt laudantium. Illo non quos eos vel ipsa. Explicabo itaque est optio neque rerum provident enim qui sed. Corrupti commodi voluptatem vero soluta hic.
              </p>
              <p>
                Modi corporis consectetur aliquid sit cum tenetur enim. Sed voluptatum quis voluptates laudantium incidunt laudantium. Illo non quos eos vel ipsa. Explicabo itaque est optio neque rerum provident enim qui sed. Corrupti commodi voluptatem vero soluta hic.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
